lib/OpenQA/Worker/Engines/isotovideo.pm

```diff
@@ -296,6 +296,7 @@ sub engine_workit {
     my $assetkeys = detect_asset_keys(\%vars);
     if (my $cache_dir = $global_settings->{CACHEDIRECTORY}) {
         $shared_cache = do_asset_caching($job, \%vars, $cache_dir, $assetkeys, $webui_host, $pooldir);
+        return $shared_cache if (ref($shared_cache) eq 'HASH' && $shared_cache->{error});
```

Comment:
Maybe you don't even need to check for `ref($shared_cache) eq 'HASH'` but just rely on the second condition

Reply:
Seems like `do_asset_caching` can return `undef`. To me it is not clear whether this is actually an error or not as well. Maybe it makes sense to adjust `do_asset_caching` to always return a meaningful error message.

Reply:
@Martchus sometimes it returns a `dir`.  Do you mean we could return a hash like this `{error => ,  shared_dir => }` in `do_asset_caching`?

Reply:
That would at least be more readable. But so far the `do_asset_caching` function might exist without an explicit `return` statement. It would be better to have an explicit `return` in all code paths to make it clear in any case what the function returns.

Reply:
@okurz I remain the checking `ref`. Because sometimes the function return a string, if I only use `$shared_cache->{error}`, it will report a error message `Can't use string ("xxx") as a HASH ref while "strict refs"`. 
@Martchus I just added a `return undef` in `do_asset_cacheing`, is it enough?

Reply:
hm, maybe I did not yet understand it completely. But I really think checking the perl data type is not a good style. IMHO we should only check the returned values.

Reply:
If I do not miss something, the `do_asset_caching` will return 3 types of values: `return undef`, `return $shared_dir (string)`, `return {error => $error_message}`.   when it return `{error => }` we need to stop the test and return error. Although checking the perl data type is not a good idea,  I have no idea how to check the value to make it more simply.

Reply:
well, you can streamline this by replacing `return undef` into `return {error => $another_error_message}` and then use

```perl
return $shared_cache->{error} if $shared_cache->{error};
```

Reply:
Do you mean using `return {error => $another_error_message}` to replace `return undef` at the end of function `do_asset_caching` ?  But there is still a situation that `do_asset_caching` return a string. So the `if $shared_cache->{error}` will report a error message that I mentioned in above message.

Reply:
BTW `return undef` should be treated as success.

Reply:
If `return undef` should be treated as success you can't implemant the suggestion by @okurz exactly but you can still follow the idea. In my opinion it is also better to return a consistent type and then only check for the presence of the error key to determine whether an error occurred. In the good case with no additional information you could return an empty hash.